### ISSUE TITLE
Psych.to_json produces invalid json for objects inherited from ActiveRecord::Base 

### ISSUE THREAD

**Issue by ianks** [rails/rails#27218]
### Steps to reproduce

I have created an executable test for this bug: https://gist.github.com/ianks/ba1f7175754780168c2a6618f13c95ec

I did not know whether to report this to Psych or Rails, just let me know if it belongs elsewhere and I will move the issue.

### Expected behavior
Psych.to_json should emit valid JSON. Currently, not all keys are double-quoted; which although valid in JS, is not valid JSON.

### Actual behavior
It emits invalid json, which can be seen in this example:

```json
{"raw_attributes": {"id": null}, "attributes": {attributes: {"id": {name: "id", value_before_type_cast: null,
        type: {precision: null, scale: null, limit: null, range: {"begin": -2147483648,
            "end": 2147483648, "excl": true}}, original_attribute: null}}}, "new_record": true,
  "active_record_yaml_version": 1}
```
### System configuration
**Rails version**: 5.0.0

**Ruby version**: 2.3.1

**Comment by ianks** [comment]
Upon further investigation, it seems the scalar types for attributes are not marked as `quoted` (corresponding code https://github.com/ruby/psych/blob/master/lib/psych/json/yaml_events.rb#L21)

On one hand, it seems like an easy fix to mark such nodes as quoted. On the other hand, `Psych.to_json` should not allow emitting invalid JSON. Thoughts @sgrif @tenderlove ?

**Comment by maclover7** [comment]
Looks like this is fixed via c4cb6862babd2665a65056e205c2a5fd17a5d99d, which is on Rails master.

**Comment by sgrif** [comment]
It looks like the only reason it's fixed on master is because we're manually setting all of the keys. The invalid keys in your example output are all places where psych was producing the keys from instance variables, which leads me to believe this is a bug in psych not Rails.
